### PR TITLE
Convert TLS connection adapter to connection middleware

### PR DESCRIPTION
**Pull request by jkotalik**
Handles part of https://github.com/aspnet/AspNetCore/issues/4623

Pretty much is a direct port of https://github.com/aspnet/KestrelHttpServer/pull/2849 to 3.0. Honestly confused why we didn't merge it before, things overall looked good with it.

Follow up:
- Implement SslDuplexPipe : Stream, IDuplexPipe to remove allocations of two pipes and the RawStream.
- Any breaking changes to remove all adapters

Drafting to write some follow-up questions to make sure design wise everything looks good.

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs [hunk 1]
```diff
@@ -371,8 +371,12 @@ private void CloseUninitializedConnection(ConnectionAbortedException abortReason
 
             _context.ConnectionContext.Abort(abortReason);
 
-            _adaptedTransport.Input.Complete();
-            _adaptedTransport.Output.Complete();
+            // TODO this probably should check if Https is available?
+            if (_context.ConnectionAdapters.Count > 0)
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs]: @davidfowl I'm not if we need this check still?

File: src/Servers/Kestrel/Core/src/KestrelServer.cs [hunk 2]
```diff
@@ -122,6 +123,12 @@ public async Task StartAsync<TContext>(IHttpApplication<TContext> application, C
 
                 async Task OnBind(ListenOptions options)
                 {
+                    // TODO Why?
+                    if (options.ConnectionAdapters.Any(c => c.IsHttps))
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/KestrelServer.cs]: Also this, I don't think we need to check any connection adapters now.

File: src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs]: Nuke all of the connection adapter code.

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs [hunk 3]
```diff
@@ -16,31 +16,34 @@
...
+        private readonly ILogger _logger;
         private readonly X509Certificate2 _serverCertificate;
         private readonly Func<ConnectionContext, string, X509Certificate2> _serverCertificateSelector;
 
-        private readonly ILogger _logger;
-
-        public HttpsConnectionAdapter(HttpsConnectionAdapterOptions options)
-            : this(options, loggerFactory: null)
+        public HttpsConnectionMiddleware(ConnectionDelegate next, HttpsConnectionAdapterOptions options)
+          : this(next, options, loggerFactory: null)
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: Pass NullLoggerFactory.Instance

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs [hunk 4]
```diff
@@ -200,7 +198,59 @@ private async Task<IAdaptedConnection> InnerOnConnectionAsync(ConnectionAdapterC
...
+                    pipelineTask = adaptedPipeline.RunAsync(sslStream);
+
+                    await _next(context);
+
+                    await pipelineTask;
+                }
+            }
+            finally
+            {
+                if (!pipelineTask.IsCompleted)
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: Don't try to optimize.

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs [hunk 5]
```diff
@@ -200,7 +198,59 @@ private async Task<IAdaptedConnection> InnerOnConnectionAsync(ConnectionAdapterC
...
+                context.Transport = adaptedPipeline;
+
+                using (adaptedPipeline)
+                using (sslStream)
+                {
+                    pipelineTask = adaptedPipeline.RunAsync(sslStream);
+
+                    await _next(context);
+
+                    await pipelineTask;
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: Remove this. Just always await in the finally.

File: src/Servers/Kestrel/Core/src/KestrelServer.cs (hunk 2 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/KestrelServer.cs]: Nuke it.

File: src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs [hunk 6]
```diff
@@ -128,7 +137,7 @@ private async Task ReadInputAsync(Stream stream)
 
                     var result = await Input.Writer.FlushAsync();
 
-                    if (result.IsCompleted)
+                    if (result.IsCompleted || result.IsCanceled)
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: !?

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: No one calls CancelPendingWrite, right?

File: src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs [hunk 7]
```diff
@@ -145,7 +154,16 @@ private async Task ReadInputAsync(Stream stream)
                 // The application could have ended the input pipe so complete
                 // the transport pipe as well
                 _transport.Input.Complete();
+
+                // Cancel any pending reads from the application
+                Output.Reader.CancelPendingRead();
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: You just completed the input writer. What point does clearing the output reader have? Was this supposed to be CancelPendingWrite?

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs]: I think this could result in truncated output.

**Review by halter73 — DISMISSED** [review]
This is good except I'd remove the cross-cancellation

**Comment by jkotalik** [comment]
@aspnet/build thoughts?
```
  Results will be available from https://mc.dot.net/#/user/aspnetcore/pr~2Faspnet~2Faspnetcore/ci/20190611.42.1
  Sent Helix Job 6fb69598-c097-4168-87bc-5020cd2b7ee9
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : ArgumentException: Provided Job List Uri is not accessible [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Client.HelixApi.HandleFailedRequest(RestApiException ex) in /_/src/Microsoft.DotNet.Helix/Client/CSharp/HelixApi.cs:line 29 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Client.HelixApi.OnFailedRequest(RestApiException ex) in /_/src/Microsoft.DotNet.Helix/Client/CSharp/generated-code/HelixApi.cs:line 116 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Client.Job.<NewInternalAsync>d__10.MoveNext() in /_/src/Microsoft.DotNet.Helix/Client/CSharp/generated-code/Job.cs:line 273 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : --- End of stack trace from previous location where exception was thrown --- [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Client.Job.<NewAsync>d__9.MoveNext() in /_/src/Microsoft.DotNet.Helix/Client/CSharp/generated-code/Job.cs:line 190 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : --- End of stack trace from previous location where exception was thrown --- [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Client.HelixApi.<RetryAsync>d__69`1.MoveNext() [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : --- End of stack trace from previous location where exception was thrown --- [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Client.JobDefinition.<SendAsync>d__81.MoveNext() in /_/src/Microsoft.DotNet.Helix/JobSender/JobDefinition.cs:line 225 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : --- End of stack trace from previous location where exception was thrown --- [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Sdk.SendHelixJob.<ExecuteCore>d__65.MoveNext() in /_/src/Microsoft.DotNet.Helix/Sdk/SendHelixJob.cs:line 234 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : --- End of stack trace from previous location where exception was thrown --- [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :    at Microsoft.DotNet.Helix.Sdk.HelixTask.Execute() in /_/src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs:line 45 [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : RestApiException: The response contained an invalid status code 400 Bad Request [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :  [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error : Body: {"Message":"Provided Job List Uri is not accessible"} [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :  [F:\workspace\_work\1\s\eng\helix\helix.proj]
F:\workspace\_work\1\s\.nuget\packages\microsoft.dotnet.helix.sdk\2.0.0-beta.19302.2\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(51,5): error :  [F:\workspace\_work\1\s\eng\helix\helix.proj]
```

**Comment by natemcmaster** [comment]
Can you open an issue on https://github.com/dotnet/core-eng/? Looks like an outage or error within Helix

File: src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs [hunk 8]
```diff
@@ -217,64 +217,90 @@ private HttpConnectionContext CreateDerivedContext(IDuplexPipe transport)
...
-                        break;
                     case ProtocolSelectionState.Aborted:
                         break;
                 }
+            }
+
+            switch (previousState)
+            {
+                case ProtocolSelectionState.Initializing:
+                    // OnReader/WriterCompleted callbacks are not wired until after leaving the Initializing state.
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs]: We should change this comment now that it's using the ConnectionClosed token.

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs [hunk 9]
```diff
@@ -69,33 +71,58 @@ public HttpsConnectionAdapter(HttpsConnectionAdapterOptions options, ILoggerFact
             }
 
             _options = options;
-            _logger = loggerFactory?.CreateLogger<HttpsConnectionAdapter>() ?? (ILogger)NullLogger.Instance;
+            _logger = loggerFactory?.CreateLogger<HttpsConnectionMiddleware>();
         }
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: nit: newline

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs [hunk 10]
```diff
@@ -69,33 +71,58 @@ public HttpsConnectionAdapter(HttpsConnectionAdapterOptions options, ILoggerFact
...
-        private async Task<IAdaptedConnection> InnerOnConnectionAsync(ConnectionAdapterContext context)
+        private async Task InnerOnConnectionAsync(ConnectionContext context)
         {
             SslStream sslStream;
             bool certificateRequired;
             var feature = new Core.Internal.TlsConnectionFeature();
             context.Features.Set<ITlsConnectionFeature>(feature);
             context.Features.Set<ITlsHandshakeFeature>(feature);
 
+            // TODO: Handle the cases where this can be null
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: Nit: remove TODO or do it.

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs [hunk 11]
```diff
@@ -69,33 +71,58 @@ public HttpsConnectionAdapter(HttpsConnectionAdapterOptions options, ILoggerFact
...
+                pool: memoryPoolFeature.MemoryPool,
+                readerScheduler: PipeScheduler.Inline,
+                writerScheduler: PipeScheduler.Inline,
+                pauseWriterThreshold: _options.MaxOutputBufferSize ?? 0,
+                resumeWriterThreshold: _options.MaxOutputBufferSize / 2 ?? 0,
+                useSynchronizationContext: false,
+                minimumSegmentSize: memoryPoolFeature.MemoryPool.GetMinimumSegmentSize()
+            );
+
+            // TODO: eventually make SslDuplexStream : Stream, IDuplexPipe to avoid RawStream allocation and pipe allocations
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: This TODO on the other hand is fine. This seems like a fair bit of work.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs (hunk 11 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: I'll tackle them across the code base in a separate PR. 

PS: https://github.com/aspnet/AspNetCore/blob/0648edf7c2c481309272697c0e0b60e47696f9f7/src/Servers/Kestrel/Transport.Libuv/src/Internal/Networking/UvStreamHandle.cs#L47

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs (hunk 10 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: I'll tackle them across the code base in a separate PR. 

PS: https://github.com/aspnet/AspNetCore/blob/0648edf7c2c481309272697c0e0b60e47696f9f7/src/Servers/Kestrel/Transport.Libuv/src/Internal/Networking/UvStreamHandle.cs#L47

**Comment by davidfowl** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=231397

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2410
https://github.com/aspnet/AspNetCore-Internal/issues/2623
https://github.com/aspnet/AspNetCore-Internal/issues/2415

File: src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs [hunk 12]
```diff
@@ -215,64 +215,90 @@ private HttpConnectionContext CreateDerivedContext(IDuplexPipe transport)
 
         private void StopProcessingNextRequest()
         {
+            ProtocolSelectionState previousState;
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs]: Do we still need these changes?

**Review by jkotalik — COMMENTED** [review]
I approve, but this is my PR 😃

**Comment by jkotalik** [comment]
I'm going to go ahead and merge this. I'm OOF friday so I'd like to be able to revert if something is failing due to this change.

File: src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs (hunk 10 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/HttpsConnectionMiddleware.cs]: I'm aware there are TODO comments in Kestrel. That one was checked in [five years ago](https://github.com/aspnet/AspNetCore/commit/de6c32dc4b33c066bad84e65b76e05e46202769a) when Kestrel was still a test server. I bet you could find a bunch more recent examples than that, but I don't really see the point. I'm glad you don't plan to just leave this there though.

**Pull request merged**